src/fuzz-canonicalize.zig

```diff
@@ -0,0 +1,84 @@
+//! Fuzzing for the Roc canonicalization phase
+//!
+//! This fuzzer tests the canonicalization phase of the Roc compiler, which transforms
+//! the parsed AST into a canonical intermediate representation (CIR). The fuzzer is
+//! designed to bias toward inputs that successfully parse and pass through canonicalization
+//! without errors, as these provide better code coverage for finding edge cases.
+//!
+//! ## Building
+//!
+//! Note: Compiling the fuzz tests requires llvm and does not currently work in our nix shell on all systems.
+//!
+//! To build with fuzzing support:
+//!   zig build -Dfuzz
+//!
+//! To build just the repro executable (no AFL++ required):
+//!   zig build repro-canonicalize
+//!
+//! ## Running
+//!
+//! To run the fuzzer:
+//!  1. zig build -Dfuzz
+//!  2. zig build snapshot -- --fuzz-corpus /tmp/corpus  # Optional: generate corpus from snapshots
+//!  3. ./zig-out/AFLplusplus/bin/afl-fuzz -i src/fuzz-corpus/canonicalize -o /tmp/canonicalize-out/ zig-out/bin/fuzz-canonicalize
+//!
+//! To reproduce a crash:
+//!   ./zig-out/bin/repro-canonicalize /tmp/canonicalize-out/default/crashes/id:000000...
+//!   # Or with verbose output:
+//!   ./zig-out/bin/repro-canonicalize --verbose /tmp/canonicalize-out/default/crashes/id:000000...
+//!
+//! ## Corpus
+//!
+//! The initial corpus in src/fuzz-corpus/canonicalize/ contains simple valid Roc programs
+//! that pass parsing and canonicalization. This helps the fuzzer start with good examples
+//! that explore more of the canonicalization code paths rather than getting stuck on
+//! parse errors.
+//!
+//! Other afl commands also available in `./zig-out/AFLplusplus/bin`
+
+const std = @import("std");
+const coordinate_simple = @import("coordinate_simple.zig");
+
+/// Hook for AFL++ to initialize the fuzz test environment.
+pub export fn zig_fuzz_init() void {}
+
+/// Hook for AFL++ to run the fuzz test.
+pub export fn zig_fuzz_test(buf: [*]u8, len: isize) void {
+    zig_fuzz_test_inner(buf, len, false);
+}
+
+/// Inner implementation that can be called with debug flag for repro executable
+pub fn zig_fuzz_test_inner(buf: [*]u8, len: isize, debug: bool) void {
+    // We reinitialize the gpa on every loop of the fuzzer.
+    // This enables the gpa to do leak checking on each iteration.
+    var gpa_impl = std.heap.GeneralPurposeAllocator(.{}){};
+    defer {
+        _ = gpa_impl.deinit();
+    }
+    const gpa = gpa_impl.allocator();
+
+    const input = buf[0..@intCast(len)];
+
+    // Process the input through the full compiler pipeline
+    var result = coordinate_simple.processSource(gpa, input, "<fuzz>") catch |err| {
+        if (debug) {
+            std.debug.print("Processing failed with error: {}\n", .{err});
```

Comment:
So this is seen as a successful failure, correct? As in, this would be the case where there is an error and we surface it properly to the user?